### PR TITLE
libpod/Container.readFromJournal(): don't skip the first entry

### LINKED ISSUE #11253: rhel9 + log-driver=journald : missing first line
On RHEL9:
```console
# podman run --log-driver=journald  --name foo quay.io/libpod/testimage:20210610  sh -c 'echo a;echo b'
a
b
# podman logs foo
b
```
Seems to be 100% reproducible, and is the common element behind a large number of the RHEL9 gating tests.

podman-3.3.0-0.18.module+el9beta+12201+6a03aeff.x86_64

### PR DESCRIPTION
**Pull request by nalind**
When reading log entries from the journal, don't skip past the first matching entry after we've positioned the cursor at it.

Fixes #11253.

### CODE REVIEW

**Comment by vrothberg** [comment]
LGTM

I have change my opinion also since the meeting.  I think that https://github.com/containers/podman/pull/10653 would actually pass when using c/common v0.43.2.

**Comment by nalind** [comment]
Hmm, when I run tests manually in my VM, some of them are still failing.  This might require some tweaking yet.

**Comment by vrothberg** [comment]
@nalind, could you pick up https://github.com/containers/podman/pull/10653/files and bump c/common?

**Comment by nalind** [comment]
> Hmm, when I run tests manually in my VM, some of them are still failing. This might require some tweaking yet.

Oof, that wasn't with this PR, but it looked better locally once I'd checked out the pull request.

> @nalind, could you pick up https://github.com/containers/podman/pull/10653/files and bump c/common?

Okay, I think that's in here now.

**Review by vrothberg — APPROVED** [review]
LGTM

**Comment by nalind** [comment]
That extra blank entry that #10653 is adding when each container starts seems to be in conflict here.

**Comment by nalind** [comment]
Let's see what happens when we make that first entry a "history" event.

**Comment by mheon** [comment]
One failure looks like a potential network flake

**Comment by nalind** [comment]
My test system had `events_logger = "file"` set in `/usr/share/containers/containers.conf`, which was causing the `podman logs - multi journald` test to fail.

**Comment by edsantiago** [comment]
This is a real error caused by the new c/common:
```
$ podman tag 961769676411f082461f9ef46626dd7a2d1e2b2a38e6a44364bcbecf51e66dd4 quay.io/libpod/alpine@sha256:634a8f35b5f16dcf4aaa0822adc0b1964bb786fca12f6831de8ddc45e5986a00
Error: quay.io/libpod/alpine@sha256:634a8f35b5f16dcf4aaa0822adc0b1964bb786fca12f6831de8ddc45e5986a00: tag by digest not supported
```

**Comment by nalind** [comment]
> This is a real error caused by the new c/common:
> 
> ```
> $ podman tag 961769676411f082461f9ef46626dd7a2d1e2b2a38e6a44364bcbecf51e66dd4 quay.io/libpod/alpine@sha256:634a8f35b5f16dcf4aaa0822adc0b1964bb786fca12f6831de8ddc45e5986a00
> Error: quay.io/libpod/alpine@sha256:634a8f35b5f16dcf4aaa0822adc0b1964bb786fca12f6831de8ddc45e5986a00: tag by digest not supported
> ```

Oh, I don't think we're supposed to be able to "tag" images with names that don't include tags (a.k.a. tagged references).  I wrote that test, so I guess it's fitting that I get to fix it.

**Comment by TomSweeneyRedHat** [comment]
LGTM once the flake bit is hunted down.

**Review by vrothberg — COMMENTED** [review]
/lgtm
/hold

Great work, @nalind (and @rhatdan!).

**Comment by vrothberg** [comment]
The failure in the system tests seems consistent but only on `sys remote ubuntu-2104 root host`.

> [+0491s] not ok 149 podman kill - test signal handling in containers

Not sure what to do other than rebasing and trying again?

**Comment by nalind** [comment]
Rebased.

**Comment by nalind** [comment]
> Let's see what happens when we make that first entry a "history" event.

To answer my own question, the `events` logic encounters the log entry, and since we don't set the log entry's "PODMAN_TYPE" attribute to one of container/image/network/pod/system/volume, `libpod/events.newEventFromJournalEntry()` passes an empty string to `libpod/events.StringToType()`, which returns `ErrEventTypeBlank`, and `libpod/events.EventJournalD.Read()` ignores the entry.

**Comment by nalind** [comment]
The `logs: adjust handling around partial log messages` commit changes how the HTTP `logs` handler works.  The previous behavior didn't correctly handle cases where a container was outputting one character at a time (for example if the container was a shell sitting at a shell prompt and a person was typing) - the client would receive each logged character as if it were a different line of output.  But I'd appreciate a nod from people who are more familiar with it.

**Review by vrothberg — COMMENTED** [review]
LGTM

@containers/podman-maintainers PTAL on the new changes.

**Comment by vrothberg** [comment]
@Luap99 PTAL

**Comment by vrothberg** [comment]
Note: once merged, we can close https://github.com/containers/podman/issues/10587

File: pkg/api/handlers/compat/containers_logs.go
Comment (Luap99) [inline pkg/api/handlers/compat/containers_logs.go:155]: Does this match docker?

**Comment by Luap99** [comment]
This is a breaking change for API users. Therefore this should not be backported to any 3.X branches. I am fine to merge this since we are on 4.0 on main but this should be documented somewhere.

**Comment by vrothberg** [comment]
> This is a breaking change for API users. Therefore this should not be backported to any 3.X branches. I am fine to merge this since we are on 4.0 on main but this should be documented somewhere.

Unless it's a bug fix :) Let's explore what Docker does.

**Comment by Luap99** [comment]
> > This is a breaking change for API users. Therefore this should not be backported to any 3.X branches. I am fine to merge this since we are on 4.0 on main but this should be documented somewhere.
> 
> Unless it's a bug fix :) Let's explore what Docker does.

The log change is not really a bug fix. The new lines are now added on the server side and not on the client side for the libpod endpoint. Therefore this is breaking libpod api users.

File: pkg/api/handlers/compat/containers_logs.go
Comment (nalind) [reply pkg/api/handlers/compat/containers_logs.go:155]: We use the same handler for both types of clients, so this makes both behave the same.  Anecdotally, when I pointed a `docker` client at the podman socket, I get what looks like the correct behavior on the client end.

**Comment by vrothberg** [comment]
/hold cancel

Great work, @nalind and @rhatdan!

**Review by ashley-cui — COMMENTED** [review]
/lgtm

**Pull request merged**
